README
NAME
    Mojolicious::Plugin::Responsinator - Simulate screen sizes

VERSION
    0.02

DESCRIPTION
    This module allow you to embed a given web page inside an iframe to see
    how it would look on different screens.

    This is probably just a module you want to use while developing, and not
    bundle it with the final application. Example usage:

      sub startup {
        my $self = shift;
        $self->plugin("responsinator") if $ENV{ENABLE_RESPONSINATOR};
      }

SYNOPSIS
    You need to enable the plugin in your Mojolicious application:

      use Mojolicious::Lite;
      plugin "responsinator";
      get "/" => sub { shift->render(text => "test\n") };
      app->start;

    Then from the browser, you can ask for an URL with the "_size" param to
    embed a website inside an iframe. Example:

      http://localhost:3000/some/path?_size=iphone          # iphone landscape
      http://localhost:3000/some/path?_size=iphone:portrait # iphone portrait
      http://localhost:3000/some/path?_size=100x400         # width: 100px; height: 400px

PREDEFINED SIZES
    You can replace "iphone" in the example above with any of the predefined
    sizes below:

    *   iphone

    *   iphone-5

    *   ipad

    *   wildfire

    *   nexus-4

METHODS
  register
      $self->reqister($app, \%config);
      $app->plugin(responsinator => \%config);

    Will register an "around_dispatch" hook, which will trigger on the
    "_size" query param. %config can contain:

    *   param

        Use this to specify another query param than the default "_size".

COPYRIGHT
  Images
    The images are provided by the The Responsinator Team,
    <http://www.responsinator.com>.

  Code
    The code is written by Jan Henning Thorsen, <http://thorsen.pm>.

AUTHOR
    Jan Henning Thorsen - "jhthorsen@cpan.org"



Changes
Revision history for Mojolicious-Plugin-Responsinator

0.02     Mon Mar 24 11:50:32 2014
       * Fix removing default layout when rendering iframe

0.01     Thu Mar 20 22:08:11 2014
       * Started project
       * Add cool feature


lib/Mojolicious/Plugin/MediaQuery.pm
package Mojolicious::Plugin::MediaQuery;

=head1 NAME

Mojolicious::Plugin::MediaQuery - Simulate media queries

=head1 VERSION

0.01

=head1 DESCRIPTION

This module allow you to embed a given web page inside an iframe, to see
how it would look on different screens.

=head1 SYNOPSIS

You need to enable the plugin in your L<Mojolicious> application:

  use Mojolicious::Lite;
  plugin "MediaQuery";
  get "/" => sub { shift->render(text => "test\n") };
  app->start;

The from the browser, you can ask for an URL with the "_mq" param to embed a
website inside an iframe. Example:

  http://localhost:3000/some/path?_mq=iphone   # iphone portrait
  http://localhost:3000/some/path?_mq=r:iphone # iphone landscape
  http://localhost:3000/some/path?_mq=100x400  # width: 100px; height: 400px

=cut

use Mojo::Base 'Mojolicious::Plugin';

our $VERSION = '0.01';

=head1 ATTRIBUTES

=head2 presets

Defined presets:

  desktop: 992x
  ipad:    1024x768
  iphone:  320x480
  large:   1200x
  tablet:  768x

=cut

has presets => sub {
  return {
    desktop => '992x',
    ipad => '1024x768',
    iphone => '320x480',
    large => '1200x',
    tablet => '768x',
  };
};

=head1 METHODS

=head2 register

  $self->reqister($app, \%config);
  $app->plugin(MediaQuery => \%config);

Will register an "around_dispatch" hook, which will trigger on the C<_mq>
query param. C<%config> can contain:

=over 4

=item * param

Use this to specify another query param than the default "_mq".

=item * presets

This should be a hash-ref with the same format as the attribute L</presets>.

=back

=cut

sub register {
  my($self, $app, $config) = @_;
  my $param = $config->{param} || '_mq';
  my %presets = ( %{ $config->{presets} || {} }, %{ $self->presets } );

  $app->hook(around_dispatch => sub {
    my($next, $c) = @_;
    my $size = $c->param($param) or return $next->();
    my $url = $c->req->url->to_abs;
    my($rotate, $height, $width);

    $rotate = $size =~ s!^r:!!;
    $size = $presets{$size} if $presets{$size};
    $size =~ /^(\d*)x(\d*)$/ or return $next->();
    $width = $1 ? "${1}px" : "100%";
    $height = $2 ? "${2}px" : "100%";
    ($width, $height) = ($height, $width) if $rotate;
    $url->query->remove($param); # make sure it does not recurse

    $c->render(
      text => <<"      HTML",
<html style="width:100%;height:100%;padding:0;margin:0;">
<head><title>width:$width; height:$height;</title></head>
<body style="width:100%;height:100%;padding:0;margin:0;background:#333;">
<iframe style="width:$width;height:$height;background: #fff;border:0;" src="$url"></iframe>
</body>
</html>
      HTML
    );
  });
}

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut

1;


lib/Mojolicious/Plugin/Responsinator.pm
package Mojolicious::Plugin::Responsinator;

=head1 NAME

Mojolicious::Plugin::Responsinator - Simulate screen sizes

=head1 VERSION

0.02

=head1 DESCRIPTION

This module allow you to embed a given web page inside an iframe to see
how it would look on different screens.

This is probably just a module you want to use while developing, and not
bundle it with the final application. Example usage:

  sub startup {
    my $self = shift;
    $self->plugin("responsinator") if $ENV{ENABLE_RESPONSINATOR};
  }

=head1 SYNOPSIS

You need to enable the plugin in your L<Mojolicious> application:

  use Mojolicious::Lite;
  plugin "responsinator";
  get "/" => sub { shift->render(text => "test\n") };
  app->start;

Then from the browser, you can ask for an URL with the "_size" param to embed a
website inside an iframe. Example:

  http://localhost:3000/some/path?_size=iphone          # iphone landscape
  http://localhost:3000/some/path?_size=iphone:portrait # iphone portrait
  http://localhost:3000/some/path?_size=100x400         # width: 100px; height: 400px

=head1 PREDEFINED SIZES

You can replace "iphone" in the example above with any of the predefined sizes
below:

=over 4

=item * iphone

=item * iphone-5

=item * ipad

=item * wildfire

=item * nexus-4

=back

=cut

use Mojo::Base 'Mojolicious::Plugin';
use Cwd;

our $VERSION = '0.02';

=head1 METHODS

=head2 register

  $self->reqister($app, \%config);
  $app->plugin(responsinator => \%config);

Will register an "around_dispatch" hook, which will trigger on the C<_size>
query param. C<%config> can contain:

=over 4

=item * param

Use this to specify another query param than the default "_size".

=back

=cut

sub register {
  my($self, $app, $config) = @_;
  my $param_name = $config->{param} || '_size';

  push @{ $app->static->paths }, $self->_asset_path;
  push @{ $app->renderer->paths }, $self->_asset_path;

  $app->hook(around_dispatch => sub {
    my($next, $c) = @_;
    my $size = $c->param($param_name) or return $next->();
    my $frame_url = $c->req->url->to_abs;
    my $orientation = $size =~ s!:(landscape|portrait)$!! ? $1 : 'landscape';
    my @size = $size =~ /^(\d*)x(\d*)$/;

    $frame_url->query->remove($param_name); # make sure it does not recurse
    $c->render(
      layout => $config->{layout} || undef,
      template => 'responsinator',
      frame_url => $frame_url,
      identifier => $size,
      orientation => $orientation,
      param_name => $param_name,
      height => $size[1] ? "$size[1]px" : 0,
      width => $size[0] ? "$size[0]px" : 0,
    );
  });
}

sub _asset_path {
  my $asset_path = Cwd::abs_path(__FILE__);
  $asset_path =~ s!\.pm$!!;
  $asset_path;
}

=head1 COPYRIGHT

=head2 Images

The images are provided by the The Responsinator Team, L<http://www.responsinator.com>.

=head2 Code

The code is written by Jan Henning Thorsen, L<http://thorsen.pm>.

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut

1;
